[Excel] (Custom functions) Authentication article (#711)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-authentication.md

Suggested usage:

- Use `AsyncStorage` to set and store user tokens.
- From your custom function, read user tokens from `AsyncStorage` and use them to make your web request.
- For best performance, you should save a user token into a global variable within custom functions so you can access it from other functions without having to look it up in `AsyncStorage` again.

Additionally, there are some benefits to using `AsyncStorage`; it offers a sandbox environment on a user's device and cannot be accessed by other add-ins.  

## Dialog API

If a token doesn't exist, you should use the Dialog API to ask the user to sign in. Then, when a user enters their credentials into the dialog box, that information can be stored in `AsyncStorage`.

> Keep in mind that both the task pane (and other UI elements of an add-in) and custom functions have their own separate Dialog APIs. They're both referred to as the "Dialog API", but use `Officeruntime.Dialog` to authenticate users in the custom functions runtime.

Reviewer: I would phrase this a bit more specific and link to both API references if they want to see the nuanced details.
```suggestion
NOTE: The custom functions runtime uses a Dialog object that is slightly different from the Dialog object in the runtime used by task panes. They're both referred to as the "Dialog API", but use `Officeruntime.Dialog` to authenticate users in the custom functions runtime.
```

Reviewer: Seems like this might work as a NOTE as well.

> For information on how to use the `OfficeRuntime.Dialog`, consult [Custom Functions runtime](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-runtime?view=office-js#displaying-a-dialog-box).

Reviewer 2: General MS style: For more information..., see...

When envisioning the entire authentication process as a whole, it might be helpful to think of `OfficeRuntime`, the task pane and UI elements of your add-in, and the custom functions portions of your add-in as separate entities which can communicate with each other. As shown in the following diagram, calls made from a cell in an Excel workbook and `OfficeRuntime.Dialog` exist in the custom functions runtime, indicated in orange. They communicate back and forth to the task pane through an intermediary, `OfficeRuntime.AsyncStorage`.

> ![Diagram of custom functions, OfficeRuntime, and taskpanes working together.](../images/RuntimeDiagram.svg "Authentication diagram.")

Reviewer 2: Can we make the line from the dialog straight, and extend the asyncstorage box to the left?

Author: Yeah definitely possible. What do you think about the usage of color? Maybe I can make it have a large border or something so that color isn't the only demarcation

## General guidance

> Use any authentication method you prefer. There is no particular pattern or method you must follow to implement your own authentication with custom functions. You may wish to consult the documentation about various authentication patterns, starting with [this article about authorizing via external services](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/auth-external-add-ins?view=office-js).  

Reviewer: ```suggestion
Office Add-ins are web-based and you can use any web authentication technique. There is no particular pattern or method you must follow to implement your own authentication with custom functions. You may wish to consult the documentation about various authentication patterns, starting with [this article about authorizing via external services](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/auth-external-add-ins?view=office-js).  
```

> There are two locations which should not be used to store data if you're using custom functions:  

Reviewer: ```suggestion
Avoid using the following locations to store data when developing custom functions:  
```

---
ms.date: 1/28/2019

> description: Learn how to authenticate users via custom functions in Excel.

Author: Makes sense about via, didn't even think about that!

title: Authentication for Custom Functions
---

# Authentication

> You may wish to verify that a user is authenticated before they access and use your custom function. While custom functions does not require a specific method of authentication, you should be aware that custom functions runs in a separate runtime from the task pane and other UI elements of your add-in. Because of this, you will need to pass data back and forth between the two runtimes using `AsyncStorage` and the Dialog API.

Reviewer 3: does not > doesn't

  
## AsyncStorage

> Custom functions' runtime does not have a `localStorage` object available on the global window, where you might typically store user information. Instead, custom functions and task panes both have access to another storage location, `OfficeRuntime.AsyncStorage`. It has a variety of [useful properties](https://docs.microsoft.com/en-us/javascript/api/office-runtime/officeruntime.asyncstorage?view=office-js) to set and store data. It essentially serves as a bridge between custom functions and your task pane (or other UI elements).

Reviewer 3: Essentially, 'AsyncStorage' serves as a bridge between custom functions and your task pane (or other UI elements)

Suggested usage:

- Use `AsyncStorage` to set and store user tokens.

> - From your custom function, read that token from `AsyncStorage` and use it to make your web request.

Reviewer 3: For consistency, say "From your custom function, read user tokens from 'AsyncStorage' and use them to make web requests.

> - For best performance, you should save this token into a global variable within custom functions so you can access it from other functions without having to look it up in `AsyncStorage` again.

Reviewer 3: "you should save a user token into a global variable" (Instead of "this token" and "that token."

Additionally, there are some benefits to using `AsyncStorage`; it offers a sandbox environment on a user's device and cannot be accessed by other add-ins.  

## Dialog API

If a token doesn't exist, you should use the Dialog API to ask the user to sign in. Then, when a user enters their credentials into the dialog box, that information can be stored in `AsyncStorage`.

> Keep in mind that both the task pane (and other UI elements of an add-in) and custom functions have their own, separate Dialog APIs. They are both referred to as the "Dialog API", but you will want to use `Officeruntime.Dialog` to authenticate users as this is the dialog that works in the custom functions runtime.

Reviewer 3: This is a little wordy, you can try something a little more straightforward:
They are both referred to as the "Dialog API", but use `Officeruntime.Dialog` to authenticate users in the custom functions runtime.

For information on how to use the `OfficeRuntime.Dialog`, consult [Custom Functions runtime](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-runtime?view=office-js#displaying-a-dialog-box).

> When envisioning the entire authentication process as a whole, it might be helpful to think of `OfficeRuntime`, the task pane and UI elements of your add-in, and the custom functions portions of your add-in as separate entities which can communicate with each other. As shown in the following diagram, calls made from a cell in an Excel workbook and `OfficeRuntime.Dialog` exist in the custom functions runtime, indicated in orange. They communicate back and forth to the task pane through an intermediary, `OfficeRuntime.AsyncStorage`.

Author: @davidchesnut Ideas? Maybe we can brainstorm on this diagram. I threw it in there because I felt like without a diagram it's totally confusing, but maybe we can rely on people to have read the TBD article about architecture?

> When envisioning the entire authentication process as a whole, it might be helpful to think of `OfficeRuntime`, the task pane and UI elements of your add-in, and the custom functions portions of your add-in as separate entities which can communicate with each other. As shown in the following diagram, calls made from a cell in an Excel workbook and `OfficeRuntime.Dialog` exist in the custom functions runtime, indicated in orange. They communicate back and forth to the task pane through an intermediary, `OfficeRuntime.AsyncStorage`.

Reviewer: I updated the diagram with colors that should work for now. I'm submitting an art request so we can have an actual artist make it look better later.

![Diagram of custom functions, OfficeRuntime, and taskpanes working together.](../images/RuntimeDiagram.svg "Authentication diagram.")

## General guidance

> Use any authentication method you prefer. There is no particular pattern or method you must follow to implement your own authentication with custom functions. You may wish to consult the documentation about various authentication patterns; you may want to start with [this article about authorizing via external services](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/auth-external-add-ins?view=office-js).  

Reviewer 3: You may wish to consult the documentation about various authentication patterns starting with [this article...

Author: Is "you may wish to consult" too formal sounding? I wasn't sure how prescriptive we can be...we could change to something like "To get started with authentication patterns, a good start is this article about blah"

Reviewer 3: Yeah, it isn't very Microsoft voice-y! Your second choice is much better!

> There are some locations which should not be used to store data if you are using custom functions:  

Reviewer 3: "There are two locations you should not use to store data if you are using custom functions: " (Saying "some locations" implies there might be more.)

> - `localStorage`: Custom functions do not have access to the global `window` object and therefore have no access to data stored in `localStorage`.

Reviewer 3: Bullet these two lines
